Use Typhoeus for more performant CDN requests

igor-makarov opened the pull request:
This PR comes to address 3 issues regarding `CDNSource` and its HTTP handling by moving to `Typhoeus`, a `libcurl` wrapper:

### 50 Threads
The previous implementation utilized 50 threads to make HTTP requests, which was too much/too little for different use cases.  
The new implementation moves all the async code to a single-threaded reactor provided by the [libcurl Multi interface](https://curl.haxx.se/libcurl/c/libcurl-multi.html). This achieves same performance as the multithreaded way.

### No HTTP/2
The built-in `Net::HTTP` module doesn't support HTTP/2, and by extension, `nap` also doesn't.  There are many benefits to HTTP/2, but the relevant to us are reducing the number of connections and reducing bandwidth by compressing headers.

I believe that the time to move to this better-performing implementation is now, before 1.8.0 ships to the general public and we start seeing much more traffic from default-repo users.

### Gzip
`nap` doesn't have built-in support for Gzip decoding the requests. Typhoeus does and now so does `CDNSource`.

dnkoutso commented:
Curious about some before/after stats here, what makes it faster/more performant? Also recommend this for 1.9 as it seems a bigger change for 1.8 which is already in beta and things are working, i.e no major issues reported.

igor-makarov commented:
The main performance gain here is the use of HTTP/2 and not requiring tons of threads.

The reason I'd like to get it into 1.8.0 (and the Netlify redirect on Specs) is so we don't get any nasty surprises when everyone switches over to default CDN and starts hammering GitHub's HTTP 1.0 servers with 50 requests per client.

igor-makarov commented:
@dnkoutso please review! 💪

dnkoutso commented:
Talked offline we will push this to 1.9.

dnkoutso commented:
lets point to master and rebase so we can land this.

igor-makarov commented:
Rebased, also added `.netrc` support for free! 💪  
@igorudovika this means #588 isn't necessary.

However, there's still the `HTTP` class that doesn't use Typhoeus.  
We'll need to transition it if we want `.netrc` support across the board.

I'm thinking that we either need to add a Typhoeus extension method and use it in all cases or add the CDN client use cases to `HTTP`.

igorudovika commented:
> #588

Thanks @igor-makarov  I thought to do similar change in  #588 after this one merge but if it is done here seems you are right  #588 is not mandatory. 
Can I help you in order to fix `http` class? I can change  #588 to fix only this left over.

igor-makarov commented:
I think it'll be better if `Http` switches to using Typhoeus, I'll do it after this PR lands if that's alright.
@dnkoutso what do you think?

igorudovika commented:
@igor-makarov I have one more question regarding Typhoeus. In current 1.7.x release REST from Fingertips/nap works fine with self-signed SSL certificates. Perhaps is not intentionally but it is. But with current implementation seems it is not true. It may be problem for mine task(s). I do not see any way to make self-signed SSL working without code change after this PR will be merged. Do you have any suggestions? 

For POD library fetching I can use .curlrc configuration file. Cocoapods client uses external  curl bin to get them by http(s) protocol. But how to handle Specs and CDN metadata it is not clear.

igor-makarov commented:
I see what you mean. Nap is an old library, and SSL certificate verification is opt-in there. 

cURL and libcurl are continuously updated, and SSL verification is on by default. To disable it, you'll need to pass `:ssl_verifypeer => false` and `:ssl_verifyhost => 0` to Typhoeus.

@dnkoutso I'm beginning to get a feeling of a scope creep here. Do you think we should do all this authentication work for CDN, or perhaps this is beyond the scope of the CocoaPods CDN client?

igorudovika commented:
@igor-makarov Yep, I was talking about `:ssl_verifypeer => false and :ssl_verifyhost => 0`
I understand this more work for you and ready to help. I can create PR with such functionality and add work with Typhoeus in `http` class(or refactor in such way  #588)

BTW: very small change 19eafe4e39e78888741e44f689c2484751b1b5fa allows to use http in CDN client against https. It can be useful for some users and workaround for this issue. I just have tested it. Works fine

igorudovika commented:
I mean commit https://github.com/CocoaPods/CocoaPods/pull/9164/commits/19eafe4e39e78888741e44f689c2484751b1b5fa

dnkoutso commented:
@igor-makarov one more rebase and I merge this!

igor-makarov commented:
Rebased! 💪🏻

igor-makarov commented on cocoapods-core.gemspec:
```diff
@@ -24,6 +24,8 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency 'fuzzy_match', '~> 2.0.4'
   s.add_runtime_dependency 'algoliasearch', '~> 1.0'
   s.add_runtime_dependency 'concurrent-ruby', '~> 1.1'
```
Yeah, definitely. I've wrapped the Typhoeus callbacks in Promises so that waiting on concurrent requests is very idiomatic and clearly represented.

igor-makarov commented:
🎉
